third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -146,19 +154,40 @@ impl Interpreter {
         }
 
         let mut current_frame = self
-            .make_new_frame(gas_meter, loader, function, ty_args, locals)
+            .make_new_frame(
+                gas_meter,
+                entrypoint_resolver.loader(),
+                function,
+                ty_args,
+                locals,
+            )
             .map_err(|err| self.set_location(err))?;
 
         // Access control for the new frame.
         self.access_control
             .enter_function(&current_frame, current_frame.function.as_ref())
             .map_err(|e| self.set_location(e))?;
+
         loop {
-            let resolver = current_frame.resolver(loader, module_store);
-            let exit_code =
-                current_frame //self
-                    .execute_code(&resolver, &mut self, data_store, module_store, gas_meter)
-                    .map_err(|err| self.attach_state_if_invariant_violation(err, &current_frame))?;
+            // Select which resolver to use: if we are executing in the context of the entrypoint
+            // function, we use the one provided by the caller.
+            let resolver = if self.call_stack.is_empty() {
+                BorrowedOrOwnedResolver::Borrowed(entrypoint_resolver)
```

Comment:
For CoW one need to implement ToOwned, which here we do not need because we only care about referencing, and ensuring the owned/borrowed versions have same lifetimes, but without copy. Not sure if rust has CoW-like data structure without cloning, or if there is a trick around?

Reply:
Found https://crates.io/crates/maybe-owned, do we want to use an external library for that? Though it is not so difficult to have our own generic aptos implementation, what do you think? For StorageAdapter / resolvers o think we also have a problem that we pass a reference to state view, but sometimes we could have owned state view and this makes code ugly with both state view, and the resolver, alive at the same time. They seem like good candidates for this pattern as well?